op-e2e/interop/system2.go

```diff
@@ -0,0 +1,408 @@
+		PollInterval:             50 * time.Millisecond,
+		TxMgrConfig:              setuputils.NewTxMgrConfig(s.l1.UserRPC(), batcherSecret),
+		LogConfig: oplog.CLIConfig{
+			Level:  log.LevelInfo,
+			Format: oplog.FormatText,
+		},
+		Stopped:               false,
+		BatchType:             derive.SpanBatchType,
+		MaxBlocksPerSpanBatch: 10,
+		DataAvailabilityType:  batcherFlags.CalldataType,
+		CompressionAlgo:       derive.Brotli,
+	}
+	batcher, err := bss.BatcherServiceFromCLIConfig(
+		context.Background(), "0.0.1", batcherCLIConfig,
+		logger.New("service", "batcher"))
+	require.NoError(t, err)
+	require.NoError(t, err, "must start batcher")
+	require.NoError(t, batcher.Start(context.Background()))
+	t.Cleanup(func() {
+		ctx, cancel := context.WithCancel(context.Background())
+		cancel() // force-quit
+		_ = batcher.Stop(ctx)
+	})
+
+	return l2Set{
+		opNode,
+		l2Geth,
+		proposer,
+		batcher,
+		secrets}
+}
+
+// prepareL2s creates the L2s for the system, returning a map of L2s
+func (s *system2) prepareL2s(t *testing.T) map[string]l2Set {
+	l2s := make(map[string]l2Set)
+	for id, l2Out := range s.worldOutput.L2s {
+		l2s[id] = s.prepareL2(t, id, l2Out)
+	}
+	return l2s
+}
+
+// addL2 adds an L2 to the system by creating the resources for it
+// and then assigning them to the system's map of L2s
+func (s *system2) addL2(t *testing.T, id string, output *interopgen.L2Output) {
+	if s.l2s == nil {
+		s.l2s = make(map[string]l2Set)
+	}
+	s.l2s[id] = s.prepareL2(t, id, output)
+}
+
+// prepare sets up the system for testing
+// it is a monlith that creates all the resources by calling the other prepare functions
+func (s *system2) prepare(t *testing.T) {
+
+	s.logger = testlog.Logger(t, log.LevelInfo)
+
+	s.hdWallet = s.prepareHDWallet(t)
+
+	s.worldDeployment, s.worldOutput = s.prepareWorld(t)
+
+	s.beacon, s.l1 = s.prepareL1(t)
+
+	s.l2s = s.prepareL2s(t)
+
+	// TODO op-supervisor
+}
+
+// this function taken from System
+func (s *system2) NodeClient(t *testing.T, name string, rpcEndpoint endpoint.RPC) *ethclient.Client {
+	nodeClient, ok := s.clients[name]
+	if ok {
+		return nodeClient
+	}
+	rpcCl := endpoint.DialRPC(
+		endpoint.PreferAnyRPC,
+		rpcEndpoint,
+		func(v string) *rpc.Client {
+			logger := testlog.Logger(t, log.LevelInfo).New("node", name)
+			cl, err := dial.DialRPCClientWithTimeout(context.Background(), 30*time.Second, logger, v)
+			require.NoError(t, err, "failed to dial eth node instance %s", name)
+			return cl
+		})
+	nodeClient = ethclient.NewClient(rpcCl)
+	s.addClient(name, nodeClient)
+	return nodeClient
+}
+
+func (sys *system2) addClient(name string, client *ethclient.Client) {
+	if sys.clients == nil {
+		sys.clients = make(map[string]*ethclient.Client)
+	}
+	sys.clients[name] = client
+}
+
+// gettter functions for the individual L2s
+// TODO: maybe the caller is just better off using the map directly, it reads nicely
+func (s *system2) getBatcher(id string) *bss.BatcherService {
+	return s.l2s[id].batcher
```

Comment:
I disagree about the caller using a map; an explicit public API, implementing an interface, is what I'd like to see long-term.

 The backend (`system2`) ideally implements an interface something like:
```
type Backend interface {
  Batcher(id ChainID) Batcher
  Proposer(id ChainID) Proposer
  // .. more services
}
```
And then `Batcher` is an interface with some helper methods, e.g. to get an RPC client binding instance.
That way we don't enshrine the Go implementations too much, and can run tests against alternative backends (Kurtosis in the future, hopefully).

And with an an internal `require.Contains` check or something, to check if the service exists, instead of running into a panic / returning nil.